src/core/public/context/context.ts

```diff
@@ -0,0 +1,172 @@
+   * key.
+   *
+   * @param contextName - The key of the {@link TContext} object this provider supplies the value for.
+   * @param provider - A {@link ContextProvider} to be called each time a new context is created.
+   * @param plugin - The plugin this provider is associated with. If `undefined`, provider gets access to all provided
+   *                 context keys.
+   * @returns The `ContextContainer` for method chaining.
+   */
+  register<TContextName extends keyof TContext>(
+    contextName: TContextName,
+    provider: ContextProvider<TContext, TContextName, TProviderParameters>,
+    plugin?: PluginName
+  ): this;
+
+  /**
+   * Create a new context.
+   *
+   * @param plugin - The plugin the context will be exposed to.
+   * @param baseContext - The initial context for the given handler.
+   * @param contextArgs - Additional parameters to call providers with.
+   * @returns A Promise for the new context object.
+   */
+  createContext(
+    plugin: PluginName,
+    baseContext: Partial<TContext>,
+    ...contextArgs: TProviderParameters
+  ): Promise<TContext>;
+}
+
+/** @internal */
+export class ContextContainerImplementation<
+  TContext extends {},
+  TProviderParameters extends any[] = []
+> implements ContextContainer<TContext, TProviderParameters> {
+  /**
+   * Used to map contexts to their providers and associated plugin. In registration order which is tightly coupled to
+   * plugin load order.
+   */
+  private readonly contextProviders = new Map<
+    keyof TContext,
+    {
+      provider: ContextProvider<TContext, keyof TContext, TProviderParameters>;
+      plugin?: PluginName;
+    }
+  >();
+  /** Used to keep track of which plugins registered which contexts for dependency resolution. */
+  private readonly contextNamesByPlugin = new Map<PluginName, Array<keyof TContext>>();
+
+  /**
+   * @param pluginDependencies - A map of plugins to an array of their dependencies.
+   */
+  constructor(private readonly pluginDependencies: ReadonlyMap<PluginName, PluginName[]>) {}
+
+  public register<TContextName extends keyof TContext>(
+    contextName: TContextName,
+    provider: ContextProvider<TContext, TContextName, TProviderParameters>,
+    plugin?: PluginName
+  ): this {
+    if (this.contextProviders.has(contextName)) {
+      throw new Error(`Context provider for ${contextName} has already been registered.`);
+    }
+
+    this.contextProviders.set(contextName, { provider, plugin });
+
+    if (plugin) {
+      this.contextNamesByPlugin.set(plugin, [
+        ...(this.contextNamesByPlugin.get(plugin) || []),
+        contextName,
+      ]);
+    }
+
+    return this;
+  }
+
+  public async createContext(
+    pluginName: PluginName,
+    baseContext: Partial<TContext> = {},
+    ...contextArgs: TProviderParameters
+  ): Promise<TContext> {
+    const contextNamesForPlugin = (plug: PluginName): Array<keyof TContext> => {
+      const pluginDeps = this.pluginDependencies.get(plug)!;
```

Comment:
I don't think it's safe to ignore the possibly `undefined` return of `this.pluginDependencies.get(...)` here. It's possible to call createContext with a `pluginName` that isn't in the `pluginDependencies`. We should probably add a defensive check and throw a descriptive error if that happens.